[FLINK-3332] Add Exactly-Once Cassandra connector

zentol opened the pull request:
2nd attempt to add an Exactly-Once Cassandra Sink. I've addressed all issues brought up in the last PR bar one: This sink only works with Tuples. A simpler Cassandra Sink is in the works (see FLINK-3311) that will feature POJO support and i intend to copy the code-paste that code once it's done.

The Exactly-once guarantee is made by saving incoming records in the OperatorState, and only committing them into Cassandra when a checkpoint completes. Whether a operator committed data is saved using a new CheckpointCommitter object, that saves this information in an external and retry-persistent resource. Note that a job failure while the data is being committed will cause duplicate data to be committed, but the chance of this happening is much smaller than for a naive At-Least-once implementation.

The CassandraExactlyOnceSink is implemented as a custom operator to get access to the Statebackend. Values are committed with single inserts using a PreparedStatement that is supplied by the user, similiar to the Batch JDBC-Outputformat.

The Exactly-Once logic is completely contained in a GenericExactlyOnceSink class that can be used by virtually every sink, requiring no knowledge about the checkpointing mechamism.

The CassandraExactlyOnceSink and GenericExactlyOnceSink are covered by tests that use the OneInputStreamTaskHarness to generate a task environment, verifying that stored data is discarded when a state is restored; all data is being committed when a notify is missed; and of course that everything works when nothing goes wrong.

uce commented:
Great addition. But in my personal opinion it's not a good idea to name the sink exactly once when you say: *Note that a job failure while the data is being committed will cause duplicate data to be committed, but the chance of this happening is much smaller than for a naive At-Least-once implementation.*

What do others sink (I'll leave this typo... :smile:)?

tillrohrmann commented:
I agree with @uce on this one. It seems to me a bit too much like marketing since, technically speaking, it gives you at least once guarantees.

tillrohrmann commented on docs/apis/streaming/connectors/cassandra.md:
```diff
@@ -0,0 +1,100 @@
+This connector provides a sink that writes data into a [Cassandra](https://cassandra.apache.org/) database.
+
+The Flink Cassandra sink integrates with Flink's checkpointing mechanism to provide
+exactly-once processing semantics. To achieve that, Flink buffers incoming records
```
Strictly speaking, this should be at least once guarantees.

zentol replied:
i am aware that there are several places where exactly-once is mentioned. once we have a consensus on this issue i will go through the PR and change all references.

tillrohrmann commented on docs/apis/streaming/connectors/cassandra.md:
```diff
@@ -0,0 +1,100 @@
+3. Table name
+
+The CassandraExactlyOnceSink is implemented as a custom operator
+instead of a sink, and as such is a bit more unwieldy than other sinks.
```
Is this paragraph necessary?

zentol replied:
hmm we could omit the second portion, but explaining why it doesn't work with addSink() is worthwhile imo.

tillrohrmann replied:
Yes delete the last subclause.

tillrohrmann commented on docs/apis/streaming/fault_tolerance.md:
```diff
@@ -176,6 +176,11 @@ state updates) of Flink coupled with bundled sinks:
     </tr>
     <tr>
+        <td>Cassandra sink</td>
+        <td>exactly once</td>
```
Sure that this is true?

rmetzger replied:
I would try to summarize the guarantees here. Or maybe add an asterisk to the "exactly once", stating that Cassandra has no support for large atomic inserts, but as long as cassandra is not failing, Flink failures will not cause duplicates.

tillrohrmann replied:
That's what is defined as at least once.

tillrohrmann replied:
I mean an at most once system will give you also exactly once processing guarantees under the assumption that nothing fails.

zentol replied:
which is also not true. a flink failure while data is being written to cassandra will cause duplicates. you can only say this if writing the data to the final table is completely handled by cassandra (for example by writing into a temporary table, exporting it to csv and importing into the target table; the only way for duplicates is if cassandra fails while importing).

tillrohrmann commented on flink-streaming-java/pom.xml:
```diff
@@ -108,6 +108,7 @@ under the License.
 				<artifactId>maven-surefire-plugin</artifactId>
 				<configuration>
 					<reuseForks>false</reuseForks>
+					<argLine>-XX:-UseSplitVerifier</argLine>
```
What is that used for?

zentol replied:
there is currently a problem with the tests that i assume are caused by powermock. i found this as a solution to the issue, although other problems still remain with the tests atm...

tillrohrmann replied:
This seems somehow wrong to me. Would be good to get the tests working without changing other modules.

zentol replied:
i agree, trying some things at the moment.

tillrohrmann commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericExactlyOnceSink.java:
```diff
@@ -0,0 +1,195 @@
+	public void processElement(StreamRecord<IN> element) throws Exception {
+		IN value = element.getValue();
+		if (serializer == null) {
+			typeInfo = TypeExtractor.getForObject(value);
```
This won't work here. Assume that you have a stream of some super class `S`. The first element you see is a sub class `A` so you will instantiate a serializer of this type. But when you see next another sub class `B`, then it will fail.

I think you should pass the serializer to the operator when it is instantiated on the client side.

zentol replied:
can i use getOperatorConfig().getTypeSerializerIn1(...) for that?

tillrohrmann replied:
I don't think so, unless you create and set the operator config somewhere.

tillrohrmann replied:
You can get the input type information from the input stream

zentol replied:
which input stream?

tillrohrmann replied:
the one you're appending the sink to

zentol replied:
so we're once again at a point where the user has to pass input TypeInformation manually, i thought we didn't want that?

tillrohrmann commented on docs/apis/streaming/connectors/cassandra.md:
```diff
@@ -0,0 +1,100 @@
+{% highlight java %}
+stream.transform(
+  "Cassandra Sink",
+  TypeExtractor.getForObject(new Tuple2<>("", 0)),
```
Why do you specify an output type for a sink?

zentol replied:
because the compiler complaints if i don' pass a typeInformation?

zentol replied:
seems like you can pass  null, would have expected some sanity check to catch that.

tillrohrmann commented:
Thanks for your contribution @zentol. Changes look good to me. I only had a comment concerning the type serializer and our usage of the term "exactly once".

rmetzger commented on docs/apis/streaming/connectors/cassandra.md:
```diff
@@ -0,0 +1,100 @@
+
+Note that the streaming connectors are currently not part of the binary distribution. See how to link with them for cluster execution [here]({{ site.baseurl}}/apis/cluster_execution.html#linking-with-modules-not-contained-in-the-binary-distribution).
+
+#### Installing RabbitMQ
```
Installing RabbitMQ?

zentol replied:
whoops.

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraExactlyOnceSink.java:
```diff
@@ -0,0 +1,113 @@
+			//insert values and send to cassandra
+			ResultSetFuture result = session.executeAsync(preparedStatement.bind(fields));
+			//add callback to detect errors
+			Futures.addCallback(result, new FutureCallback<ResultSet>() {
```
I think you can create the callback once in the open method and then pass the instance to all async calls.
This way, you save a lot of instance creations

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraExactlyOnceSink.java:
```diff
@@ -0,0 +1,113 @@
+	protected void sendValue(Iterable<IN> values) throws Exception {
+		//verify that no query failed until now
+		if (exception != null) {
+			throw new Exception(exception);
```
This will leave the cluster and session open.

zentol replied:
why is close() not called for a failing operation?

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraExactlyOnceSink.java:
```diff
@@ -0,0 +1,113 @@
+		cluster = Cluster.builder().addContactPoint(host).build();
+		session = cluster.connect();
+		if (createQuery != null) {
+			session.execute(createQuery);
```
Can the execute() throw an exception? If so, it leaves session and cluster open.

zentol replied:
does that mean that close() is not called when an operation fails in open()?

rmetzger replied:
You are right. Close is called when open fails

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraCommitter.java:
```diff
@@ -0,0 +1,77 @@
+ *
+ * Entries are in the form |operator_id | subtask_id | last_completed_checkpoint|
+ */
+public class CassandraCommitter extends CheckpointCommitter {
```
I think the entire class should use try {} finally{} blocks to properly close connections on failures.

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraCommitter.java:
```diff
@@ -0,0 +1,77 @@
+	public void commitCheckpoint(long checkpointID) {
+		SimpleStatement s = new SimpleStatement("UPDATE " + keyspace + "." + table + " set checkpoint_id=" + checkpointID + " where sink_id='" + operatorId + "' and sub_id=" + subtaskId + ";");
+		s.setConsistencyLevel(ConsistencyLevel.ALL);
+		session.executeAsync(s);
```
Why do you execute this asynchronously ? Doesn't this mean we'll never learn about errors?

zentol replied:
good point.

rmetzger commented on flink-streaming-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/streaming/connectors/cassandra/CassandraExactlyOnceSink.java:
```diff
@@ -0,0 +1,113 @@
+	public void close() throws Exception {
+		super.close();
+		session.close();
+		cluster.close();
```
I would add null checks for session and cluster

zentol replied:
I'd just add a try{...}catch{Exception e} block and log that stuff.

zentol commented:
resources are properly closed now; removed all mentions of exactly-once; tests run now that i removed powermock (and the initual cause: a final flag in ResultPartitionWriter), typeinformation is not serialized and the the serializer passed to the sink constructor.

i think i addressed all comments.

spdrnl commented:
In some circumstances it is possible to make writes to Cassandra idempotent, by adding an unique id to the insert. See the example below. Counters are probably different in this though.

```
cqlsh:mykeyspace> CREATE TABLE IF NOT EXISTS timeline (         
eid uuid,         
timesegment bigint,         
from_url text,         
to_url  text, 
PRIMARY KEY ((timesegment) , eid));

cqlsh:mykeyspace> INSERT INTO timeline (eid, timesegment, from_url, to_url) VALUES(50554d6e-29bb-11e5-b345-feff819cdc9f, 1, 'a', 'b') ;
cqlsh:mykeyspace> SELECT * FROM timeline;

 timesegment | eid                                  | from_url | to_url
-------------+--------------------------------------+----------+--------
           1 | 50554d6e-29bb-11e5-b345-feff819cdc9f |        a |      b

(1 rows)
cqlsh:mykeyspace> INSERT INTO timeline (eid, timesegment, from_url, to_url) VALUES(50554d6e-29bb-11e5-b345-feff819cdc9f, 1, 'a', 'b') ;
cqlsh:mykeyspace> SELECT * FROM timeline;
 timesegment | eid                                  | from_url | to_url
-------------+--------------------------------------+----------+--------
           1 | 50554d6e-29bb-11e5-b345-feff819cdc9f |        a |      b

(1 rows)
cqlsh:mykeyspace> INSERT INTO timeline (eid, timesegment, from_url, to_url) VALUES(uuid(), 1, 'a', 'b') ;
cqlsh:mykeyspace> SELECT * FROM timeline;

 timesegment | eid                                  | from_url | to_url
-------------+--------------------------------------+----------+--------
           1 | 50554d6e-29bb-11e5-b345-feff819cdc9f |        a |      b
           1 | 35761e5e-06b3-4ce6-9038-c9014e498281 |        a |      b
```
